File: packages/contracts/src/provenance.ts
import { createHash } from "node:crypto"

import { z } from "zod"

import { NonEmptyTextSchema, Sha256Schema, Uuid7Schema } from "./common"

export const HashPinSchema = z
  .strictObject({ ref_id: NonEmptyTextSchema, sha256: Sha256Schema })
  .readonly()

type HashPin = z.infer<typeof HashPinSchema>

function canonicalPins(pins: readonly HashPin[]): readonly HashPin[] {
  return pins.map((pin) => ({ ref_id: pin.ref_id, sha256: pin.sha256 }))
}

const PROVENANCE_MANIFEST_FIELDS = {
  source_run_id: Uuid7Schema,
  action_plan_sha256: Sha256Schema,
  code_sha256: Sha256Schema,
  environment_sha256: Sha256Schema,
  runtime_adapter_id: NonEmptyTextSchema,
  runtime_connection_id: Uuid7Schema,
  input_hashes: z.array(HashPinSchema).min(1).readonly(),
  execution_hashes: z.array(HashPinSchema).min(1).readonly(),
  output_hashes: z.array(HashPinSchema).min(1).readonly(),
  skill_hashes: z.array(HashPinSchema).min(1).readonly(),
  source_hashes: z.array(HashPinSchema).readonly(),
} as const

export const ProvenanceManifestPayloadSchema = z.strictObject(PROVENANCE_MANIFEST_FIELDS).readonly()

export type ProvenanceManifestPayload = z.infer<typeof ProvenanceManifestPayloadSchema>

export function provenanceManifestSha256(manifest: ProvenanceManifestPayload): string {
  const canonical = JSON.stringify({
    action_plan_sha256: manifest.action_plan_sha256,
    code_sha256: manifest.code_sha256,
    environment_sha256: manifest.environment_sha256,
    execution_hashes: canonicalPins(manifest.execution_hashes),
    input_hashes: canonicalPins(manifest.input_hashes),
    output_hashes: canonicalPins(manifest.output_hashes),
    runtime_adapter_id: manifest.runtime_adapter_id,
    runtime_connection_id: manifest.runtime_connection_id,
    skill_hashes: canonicalPins(manifest.skill_hashes),
    source_hashes: canonicalPins(manifest.source_hashes),
    source_run_id: manifest.source_run_id,
  }).replace(
    /[\u007f-\uffff]/g,
    (value) => `\\u${value.charCodeAt(0).toString(16).padStart(4, "0")}`,
  )
  return createHash("sha256").update(canonical, "utf8").digest("hex")
}

export const ProvenanceManifestSchema = z
  .strictObject({ ...PROVENANCE_MANIFEST_FIELDS, manifest_sha256: Sha256Schema })
  .superRefine((manifest, context) => {
    const pinGroups = [
      manifest.input_hashes,
      manifest.execution_hashes,
      manifest.output_hashes,
      manifest.skill_hashes,
      manifest.source_hashes,
    ]
    if (pinGroups.some((pins) => new Set(pins.map((pin) => pin.ref_id)).size !== pins.length)) {
      context.addIssue({
        code: "custom",
        message: "provenance ref_id values must be unique within each pin class",
      })
    }
    if (manifest.manifest_sha256 !== provenanceManifestSha256(manifest)) {
      context.addIssue({ code: "custom", message: "provenance manifest digest mismatch" })
    }
  })
  .readonly()

export type ProvenanceManifest = z.infer<typeof ProvenanceManifestSchema>

File: packages/contracts/python/science_workbench_contracts/provenance.py
import hashlib
import json
from typing import Annotated, Final, Self

from pydantic import Field, model_validator
from pydantic_core import PydanticCustomError

from .common import NonEmptyText, Sha256, Uuid7
from .task6_common import Task6ContractModel

MANIFEST_DIGEST_ERROR: Final = "provenance_manifest_digest"
MANIFEST_DIGEST_MESSAGE: Final = "provenance manifest digest mismatch"
PROVENANCE_REF_ERROR: Final = "provenance_ref_unique"
PROVENANCE_REF_MESSAGE: Final = (
    "provenance ref_id values must be unique within each pin class"
)


class HashPin(Task6ContractModel):
    ref_id: NonEmptyText
    sha256: Sha256


class ProvenanceManifest(Task6ContractModel):
    source_run_id: Uuid7
    action_plan_sha256: Sha256
    research_intent_sha256: Sha256
    code_sha256: Sha256
    environment_sha256: Sha256
    runtime_adapter_id: NonEmptyText
    runtime_connection_id: Uuid7
    input_hashes: Annotated[tuple[HashPin, ...], Field(min_length=1)]
    execution_hashes: Annotated[tuple[HashPin, ...], Field(min_length=1)]
    output_hashes: Annotated[tuple[HashPin, ...], Field(min_length=1)]
    skill_hashes: Annotated[tuple[HashPin, ...], Field(min_length=1)]
    source_hashes: tuple[HashPin, ...]
    manifest_sha256: Sha256

    @model_validator(mode="after")
    def validate_manifest_digest(self) -> Self:
        pin_groups = (
            self.input_hashes,
            self.execution_hashes,
            self.output_hashes,
            self.skill_hashes,
            self.source_hashes,
        )
        if any(len({pin.ref_id for pin in pins}) != len(pins) for pins in pin_groups):
            raise PydanticCustomError(
                PROVENANCE_REF_ERROR,
                PROVENANCE_REF_MESSAGE,
            )
        if self.manifest_sha256 != provenance_manifest_sha256(self):
            raise PydanticCustomError(
                MANIFEST_DIGEST_ERROR,
                MANIFEST_DIGEST_MESSAGE,
            )
        return self


def provenance_manifest_sha256(manifest: ProvenanceManifest) -> str:
    payload = json.dumps(
        manifest.model_dump(mode="json", exclude={"manifest_sha256"}),
        allow_nan=False,
        ensure_ascii=True,
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8")
    return hashlib.sha256(payload).hexdigest()
